scripts/docker/docker.lib.sh

```diff
@@ -0,0 +1,256 @@
+
+set -euo pipefail
+
+# A set of Docker functions written in Bash.
+#
+# Usage:
+#   $ source ./docker.lib.sh
+#
+# Arguments (provided as environment variables):
+#   DOCKER_IMAGE=ghcr.io/org/repo   # Docker image name
+#   DOCKER_TITLE="My Docker image"  # Docker image title
+
+# ==============================================================================
+# Functions to be used with custom images.
+
+# Build Docker image.
+# Arguments (provided as environment variables):
+#   dir=[path to the Dockerfile to use, default is '.']
+function docker-build() {
+
+  local dir=${dir:-$PWD}
+  _create-effective-dockerfile
+  _create-effective-version
+  docker build \
+    --progress=plain \
+    --platform linux/amd64 \
+    --build-arg IMAGE="${DOCKER_IMAGE}" \
+    --build-arg TITLE="${DOCKER_TITLE}" \
+    --build-arg DESCRIPTION="${DOCKER_TITLE}" \
+    --build-arg LICENCE=MIT \
+    --build-arg GIT_URL="$(git config --get remote.origin.url)" \
+    --build-arg GIT_BRANCH="$(_get-git-branch-name)" \
+    --build-arg GIT_COMMIT_HASH="$(git rev-parse --short HEAD)" \
+    --build-arg BUILD_DATE="$(date -u +"%Y-%m-%dT%H:%M:%S%z")" \
+    --build-arg BUILD_VERSION="$(_get-version)" \
+    --tag "${DOCKER_IMAGE}:$(_get-version)" \
+    --rm \
+    --file "${dir}/Dockerfile.effective" \
+    .
+  # Tag the image with all the stated versions
+  for version in $(_get-all-versions); do
+    docker tag "${DOCKER_IMAGE}:$(_get-version)" "${DOCKER_IMAGE}:${version}"
+  done
+  docker tag "${DOCKER_IMAGE}:$(_get-version)" "${DOCKER_IMAGE}:latest"
+  docker rmi --force "$(docker images | grep "<none>" | awk '{print $3}')" 2> /dev/null ||:
+}
+
+# Check test Docker image.
+# Arguments (provided as environment variables):
+#   args=[arguments to pass to Docker to run the container, default is none/empty]
+#   cmd=[command to pass to the container for execution, default is none/empty]
+#   dir=[path to the image directory where the Dockerfile is located, default is '.']
+#   check=[output string to search for]
+function docker-check-test() {
+
+  local dir=${dir:-$PWD}
+  # shellcheck disable=SC2086,SC2154
+  docker run --rm --platform linux/amd64 \
+    ${args:-} \
+    "${DOCKER_IMAGE}:$(_get-version)" 2>/dev/null \
+    ${cmd:-} \
+  | grep -q "${check}" && echo PASS || echo FAIL
+}
+
+# Run Docker image.
+# Arguments (provided as environment variables):
+#   args=[arguments to pass to Docker to run the container, default is none/empty]
+#   cmd=[command to pass to the container for execution, default is none/empty]
+#   dir=[path to the image directory where the Dockerfile is located, default is '.']
+function docker-run() {
+
+  local dir=${dir:-$PWD}
+  # shellcheck disable=SC2086
+  docker run --rm --platform linux/amd64 \
+    ${args:-} \
+    "${DOCKER_IMAGE}:$(_get-version)" \
+    ${cmd:-}
+}
+
+# Push Docker image.
+# Arguments (provided as environment variables):
+#   dir=[path to the image directory where the Dockerfile is located, default is '.']
```

Comment:
In Bash, this line `local dir=${dir:-$PWD}` will create a local variable called `dir` that evaluates to either the `dir` "global" variable set by the parent process or the current directory. Now, looking at this line of code, I can see why this may not be very readable. Any suggestions?

Reply:
It is used by the `_get-all-versions` function.

Reply:
To make it more explicit we could do `for version in $(dir="$dir" _get-all-versions); do`

Reply:
> To make it more explicit we could do for version in $(dir="$dir" _get-all-versions) latest; do

I like this, it makes the flow of the data more explicit.